### DISCUSSION TITLE
Why should one extend the RefType class?

### DISCUSSION THREAD

**Discussion by CaledoniaProject** [github/codeql#7695] [General]
The following class is declared to filter certain class, 

```
class Context extends RefType {
  Context() {
    this.hasQualifiedName("javax.naming", "Context")
    or
    this.hasQualifiedName("javax.naming", "InitialContext")
  }
}
```

So why does it extend the RefType class instead of any other base class?

**Comment by smowton** [comment]
It's just a type that provides `hasQualifiedName`. It would be equally valid to use `ClassOrInterface`, a subtype of `RefType` that includes other reference-types (e.g. arrays), or `Class` or `Interface` depending on which one `Context` and `InitialContext` happen to be, but `RefType` suffices so the author hasn't bothered to restrict it further.

It's worth noting that if the types `Context` and `InitialContext` happen to match the requirements of a subtype, then `Context` will be compatible with that subtype even if it doesn't `extend` it. For example, Java defines

```
class NestedType extends RefType {
  NestedType() { enclInReftype(this, _) }
  ...
}
```

Where `enclInRefType` is a database relation (something populated by the extractor) which means that `this` occurs syntactically inside some other RefType, e.g. an inner class. Supposing for a moment that `Context` really was a Java nested class, even though `Context` directly `extends RefType`, we will be able to do things like `from Context c where c.(NestedType).getEnclosingType().hasName("MyOuterClass") ...`. That is, we were able to cast `Context`, which extends `RefType`, to a `NestedType` even though it does not extend `NestedType`. The cast "worked" (that is, we got any results) because it matched `NestedType`'s characteristic predicate `enclInReftype(this, _)`.
